src/parser/uniterm.rs

```diff
@@ -305,107 +313,266 @@ impl TryFrom<UniRecord> for Types {
-/// ```
+/// We could have a simpler implementation using `std::cell::once_cell` (which [`VarKindCell`] is
+/// somehow emulating), but the latter is not stabilized yet. Used during the
+/// [FixTypeVars::fix_type_vars] phase.
+#[derive(PartialEq, Eq)]
+pub(super) struct VarKindCellData {
+    var_kind: VarKind,
+    state: VarKindCellState,
+}
+
+/// Cell providing shared mutable access to a var_kind. This is used to decide the kind of a
+/// variable associated to a forall in the `fix_type_vars` phase.
 ///
-/// Since this unirecord will eventually be interpreted as a record type, we can't know yet when
-/// parsing `fst: a` if `a` will be a type variable or a term variable (while, for all other
-/// constructs, an annotation is a boundary that `forall` binders can't cross). In this example,
-/// there is indeed an enclosing forall binding `a`. With unirecords, before fixing type variables,
-/// we have to wait until we eventually convert the unirecord to a term (in which case we fix all the
-/// top-level annotations) or a type (in which case we do nothing: the enclosing type will trigger
-/// the fix once it's fully constructed). Fixing a unirecord prior to a conversion to a term is
-/// done by [`fix_field_types`].
-pub fn fix_type_vars(ty: &mut Types) {
-    fn fix_type_vars_aux(ty: &mut Types, mut bound_vars: Cow<HashSet<Ident>>) {
-        match ty.0 {
-            AbsType::Dyn()
-            | AbsType::Num()
-            | AbsType::Bool()
-            | AbsType::Str()
-            | AbsType::Sym()
-            | AbsType::RowEmpty()
-            | AbsType::Flat(_)
-            | AbsType::Wildcard(_) => (),
-            AbsType::Arrow(ref mut s, ref mut t) => {
-                fix_type_vars_aux(s.as_mut(), Cow::Borrowed(bound_vars.as_ref()));
-                fix_type_vars_aux(t.as_mut(), bound_vars);
+/// This cell provides interior mutability for [`VarKindCellData`]. It makes it possible to mutate
+/// the inner data when put in an environment, which only provides immutable references to its
+/// values.
+#[derive(PartialEq, Eq)]
+pub(super) struct VarKindCell(RefCell<VarKindCellData>);
+
+/// Error raised by [`VarKindCell`] when trying to set a variable kind which is different from the
+/// one already set.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub(super) struct VarKindMismatch;
+
+/// Environment maintained during the `fix_type_vars` phase. Used both to determine if a variable
+/// is bound by an eclosing forall (if `env.get(var_id).is_some()`), and to provide a shared
+/// mutable variable kind that can be modified depending on the location of type variable
+/// occurrences.
+pub(super) type BoundVarEnv = Environment<Ident, VarKindCell>;
+
+impl VarKindCell {
+    /// Create a new `VarKindCell` from a mutable reference to a `VarKind`. The kind set to
+    /// `VarKind::Type` by default, meaning that unused type variables are given this kind by
+    /// default.
```

Comment:
This functions doesn't seem to take a mutable reference to a `VarKind` at all.

Reply:
Fixed 90dcc033